camayoc/data_provider.py

```diff
@@ -0,0 +1,116 @@
+        self._created_models[model.name] = model
+
+    def _create_dependencies(self, definition, new):
+        dependencies_ids = []
+        if issubclass(self._model_class, Credential):
+            return dependencies_ids
+
+        if issubclass(self._model_class, Source):
+            dependency_definitions = self._data_provider.credentials._select_definitions(
+                match_criteria={"name": Table.is_in(definition.credentials)}
+            )
+            for dependency_definition in dependency_definitions:
+                if new:
+                    dependency_definition = replace_definition_name(dependency_definition)
+                credential = self._data_provider.credentials._create_from_definition(
+                    definition=dependency_definition, new_dependencies=new, data_only=False
+                )
+                dependencies_ids.append(credential._id)
+            return dependencies_ids
+
+        if issubclass(self._model_class, Scan):
+            dependency_definitions = self._data_provider.sources._select_definitions(
+                match_criteria={"name": Table.is_in(definition.sources)}
+            )
+            for dependency_definition in dependency_definitions:
+                if new:
+                    dependency_definition = replace_definition_name(dependency_definition)
+                source = self._data_provider.sources._create_from_definition(
+                    definition=dependency_definition, new_dependencies=new, data_only=False
+                )
+                dependencies_ids.append(source._id)
+            return dependencies_ids
+
+    def _create_from_definition(self, definition, new_dependencies, data_only):
+        if existing_model := self._created_models.get(definition.name):
+            return existing_model
+
+        dependencies_ids = self._create_dependencies(definition, new=new_dependencies)
+        new_model = self._model_class.from_definition(definition, dependencies=dependencies_ids)
+
+        if not data_only:
+            self._create_model(new_model)
+
+        return new_model
+
+    def defined(self, match_criteria):
+        matching_definitions = self._select_definitions(match_criteria)
+        random.shuffle(matching_definitions)
+
+        for definition in matching_definitions:
+            model = self._create_from_definition(
+                definition=definition, new_dependencies=False, data_only=False
+            )
+            yield model
+
+    def new(self, match_criteria, new_dependencies=True, data_only=True):
+        matching_definitions = self._select_definitions(match_criteria)
+        random.shuffle(matching_definitions)
+
+        for definition in cycle(matching_definitions):
+            definition = replace_definition_name(definition)
+            model = self._create_from_definition(
+                definition=definition, new_dependencies=new_dependencies, data_only=data_only
+            )
+            yield model
```

Comment:
I was thinking about cases like [`tests.qpc.cli.test_credentials::test_clear_all`](https://github.com/quipucords/camayoc/blob/master/camayoc/tests/qpc/cli/test_credentials.py#L694), where you want multiple credentials/sources/scans. Initially I had separate methods for single object and multiple objects, but eventually decided it's easier to consider single object to be a special case - even if it's pretty common special case.

Since naming is hard, how would you approach this? I'm thinking:

```
class ModelWorker:
    def defined(self, match_criteria) -> ModelClass:
    def new(self, match_criteria) -> ModelClass:
    def defined_generator(self, match_criteria) -> Generator[ModelClass, None, None]:
    def new_generator(self, match_criteria) -> Generator[ModelClass, None, None]:
```
([Generator type signature is weird](https://peps.python.org/pep-0484/#annotating-generator-functions-and-coroutines) - TIL you can do `generator.send(value)`)

But maybe it should be `many_defined`? Or `defined_many`? Or leave `defined` as generator and add `single_defined` / `one_defined`?